registry/storage/driver/walk_test.go

```diff
@@ -45,3 +72,217 @@ func TestWalkFileRemoved(t *testing.T) {
+		// skip /folder1/file1
+		"/folder2",
+		"/folder2/file1",
+	}
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		walked = append(walked, fileInfo.Path())
+		if fileInfo.Path() == skipDir {
+			return ErrSkipDir
+		}
+		if strings.Contains(fileInfo.Path(), skipDir) {
+			t.Fatalf("skipped dir %s and should not walk %s", skipDir, fileInfo.Path())
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("expected Walk to not error %v", err)
+	}
+	compareWalked(t, expected, walked)
+}
+
+func TestWalkFallbackSkipDirOnFile(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/": {"/file1", "/file2", "/file3"},
+		},
+	}
+	skipFile := "/file2"
+	expected := []string{
+		"/file1",
+		"/file2", // return ErrSkipDir, stop early
+	}
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		walked = append(walked, fileInfo.Path())
+		if fileInfo.Path() == skipFile {
+			return ErrSkipDir
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("expected Walk to not error %v", err)
+	}
+	compareWalked(t, expected, walked)
+}
+
+// Walk is expected to skip directory on ErrSkipDir
+func TestWalkFallbackErr(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/": {"/file1", "/file2", "/file3"},
+		},
+	}
+	errFile := "/file2"
+	expected := []string{
+		"/file1",
+		"/file2", // return ErrSkipDir, stop early
+	}
+	expectedErr := errors.New("foo")
+
+	var walked []string
+	err := WalkFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		walked = append(walked, fileInfo.Path())
+		if fileInfo.Path() == errFile {
+			return expectedErr
+		}
+		return nil
+	})
+	if err != expectedErr {
+		t.Fatalf("unexpected err %v", err)
+	}
+	compareWalked(t, expected, walked)
+}
+
+// WalkFiles is expected to only walk files, not directories
+func TestWalkFilesFallback(t *testing.T) {
+	d := &fileSystem{
+		fileset: map[string][]string{
+			"/":                {"/folder1", "/file1", "/folder2"},
+			"/folder1":         {"/folder1/folder1"},
+			"/folder2":         {"/folder2/file1", "/folder2/file2"},
+			"/folder1/folder1": {"/folder1/folder1/file1", "/folder1/folder1/file2"},
+		},
+	}
+	expected := []string{
+		"/file1",
+		"/folder1/folder1/file1",
+		"/folder1/folder1/file2",
+		"/folder2/file1",
+		"/folder2/file2",
+	}
+
+	var walked []string
+	err := WalkFilesFallback(context.Background(), d, "/", func(fileInfo FileInfo) error {
+		if fileInfo.IsDir() {
+			t.Fatalf("can't walk over dir %s", fileInfo.Path())
+		}
+		if fileInfo.IsDir() != d.isDir(fileInfo.Path()) {
+			t.Fatalf("fileInfo isDir not matching file system: expected %t actual %t", d.isDir(fileInfo.Path()), fileInfo.IsDir())
+		}
+		walked = append(walked, fileInfo.Path())
+		return nil
+	})
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	compareWalked(t, expected, walked)
+}
+
+// WalkFiles is expected to stop when any error is given
```

Comment:
Looking at more of these comments . Table tests with nice descriptive name values would help do the thing you’re trying to achieve. Which I think(?) is descriptive tests that are readable and explain the goal of the test. One nice feature of table tests with name values is that when it fails, it’s easily searchable since the “name” value will show up and clearly show to the reader the intention of the test.  Very very helpful in the case of test failures.